src/ast/codegen_llvm.cpp

```diff
@@ -1672,7 +1672,10 @@ void CodegenLLVM::visit(ArrayAccess &arr)
   size_t element_size = type.GetElementTy()->GetSize();
 
   auto scoped_del_expr = accept(arr.expr);
-  array = expr_;
+  if (expr_->getType()->isPointerTy())
+    array = b_.CreatePtrToInt(expr_, b_.getInt64Ty());
```

Comment:
Why is this necessary? Doesn't `expr_` already have the right type?

Reply:
Not really, `expr_` has the right (integer) type if the array is read (proberead) from the kernel. On the other hand, if it is loaded from a map, the array has a pointer type.